python_modules/dagster/dagster/_core/definitions/resource_annotation.py

```diff
@@ -1,12 +1,52 @@
 from inspect import Parameter
 from typing import Sequence, TypeVar
 
-from typing_extensions import Annotated
+from typing_extensions import Annotated, get_args
 
 from dagster._core.decorator_utils import get_function_params
 from dagster._core.definitions.resource_definition import ResourceDefinition
 
 
+def validate_resource_annotated_function(fn) -> None:
+    """Validates any parameters on the decorated function that are annotated with
+    :py:class:`dagster.ResourceDefinition`, raising a :py:class:`dagster.DagsterInvalidDefinitionError`
+    if any are not also instances of :py:class:`dagster.ConfigurableResource` (these resources should
+    instead be wrapped in the :py:func:`dagster.Resource` Annotation).
+    """
+    from dagster import DagsterInvalidDefinitionError
+    from dagster._config.structured_config import (
+        ConfigurableResource,
+        ConfigurableResourceFactory,
+        TResValue,
+    )
+    from dagster._config.structured_config.utils import safe_is_subclass
+
+    malformed_params = [
+        param
+        for param in get_function_params(fn)
+        if safe_is_subclass(param.annotation, ResourceDefinition)
+        and not safe_is_subclass(param.annotation, ConfigurableResource)
+    ]
+    if len(malformed_params) > 0:
+        malformed_param = malformed_params[0]
+        output_type = None
+        if safe_is_subclass(malformed_param.annotation, ConfigurableResourceFactory):
+            orig_bases = getattr(malformed_param.annotation, "__orig_bases__", None)
+            output_type = get_args(orig_bases[0])[0] if orig_bases and len(orig_bases) > 0 else None
+            if output_type == TResValue:
+                output_type = None
+        raise DagsterInvalidDefinitionError(
+            """Resource param '{param_name}' is annotated as '{annotation_type}', but '{annotation_type}' outputs {value_message} value to user code such as @ops and @assets. This annotation should instead be {annotation_suggestion}""".format(
+                param_name=malformed_param.name,
+                annotation_type=malformed_param.annotation,
+                value_message=f"a '{output_type}'" if output_type else "an unknown",
+                annotation_suggestion=f"'Resource[{output_type.__name__}]'"
+                if output_type
+                else "'Resource[Any]' or 'Resource[<output type>]'",
+            )
+        )
+
+
```

Comment:
consider moving this function to `structured_config` given the all the dynamic imports